grails-app/views/stockMovement/show.gsp

```diff
@@ -75,78 +77,119 @@
+            <g:userHasRoles location="${stockMovement?.origin?.id}" roles="${[RoleType.ROLE_REQUISITION_APPROVER]}">
+                <g:set var="userHasRequestApproverRole" value="${true}"/>
+            </g:userHasRoles>
+            <g:if test="${!userHasRequestApproverRole}">
+                <g:if test="${stockMovement?.order}">
+                    <g:link controller="stockTransfer" action="edit" id="${stockMovement?.order?.id}" class="button">
+                        <img src="${resource(dir: 'images/icons/silk', file: 'pencil.png')}" />&nbsp;
+                        <warehouse:message code="default.button.edit.label" />
                     </g:link>
                 </g:if>
-                <g:elseif test="${stockMovement?.hasBeenIssued() || ((stockMovement?.hasBeenShipped() ||
-                        stockMovement?.hasBeenPartiallyReceived()) && stockMovement?.isFromOrder)}">
-                    <g:link controller="stockMovement" action="rollback" id="${stockMovement.id}" class="button">
-                        <img src="${resource(dir: 'images/icons/silk', file: 'arrow_rotate_anticlockwise.png')}" />&nbsp;
-                        <warehouse:message code="default.button.rollback.label" />
+                <g:else>
+                    <g:link controller="stockMovement" action="edit" id="${stockMovement?.id}" class="button">
+                        <img src="${resource(dir: 'images/icons/silk', file: 'pencil.png')}" />&nbsp;
+                        <warehouse:message code="default.button.edit.label" />
                     </g:link>
-                </g:elseif>
-            </g:isUserAdmin>
-                <g:set var="isPending" value="${stockMovement?.isPending() || !stockMovement?.shipment?.currentStatus}" />
-                <g:set var="originIsDepot" value="${stockMovement?.origin?.isDepot()}" />
-                <g:if test="${isPending && (isSameOrigin || !originIsDepot) && !stockMovement?.electronicType}">
-                    <g:if test="${stockMovement?.order}">
-                        <g:isUserAdmin>
-                            <g:link class="button" controller="stockTransfer" action="remove" id="${stockMovement?.id}" params="[orderId: stockMovement?.order?.id]"
-                                    onclick="return confirm('${warehouse.message(code: 'default.button.delete.confirm.message', default: 'Are you sure?')}');">
-                                <img src="${resource(dir: 'images/icons/silk', file: 'delete.png')}" />
-                                &nbsp;<warehouse:message code="default.button.delete.label" />
+                </g:else>
+                <g:link controller="partialReceiving" action="create" id="${stockMovement?.shipment?.id}" class="button">
+                    <img src="${resource(dir: 'images/icons/', file: 'handtruck.png')}" />&nbsp;
+                    <warehouse:message code="default.button.receive.label" />
+                </g:link>
+                <g:isUserAdmin>
```

Comment:
I built you up in the comment above for assigning it to a variable, but here I can see that you call the `<g:isUserAdmin>` a few times - maybe it'd be good from the performance POV to do it also in this place?

btw... I don't know if it'd be a good direction in the future, but why can't we just return user roles in the `model` while rendering the view?... it's a "loud thinking", but I'll leave this comment here in case anyone wanted to comment on that.

Reply:
this is old code whcih github highlighted as a new change since I wrapped a  `<g:if`> and `<g:userHasRoles>` around it

Reply:
Can someone add to the tech debt backlog an investigation into the right way to do these types of tag libs? It should be under Security and Performance. We invoke things like "isInRole" all the time throughout the code and I know it degrades performance. But what I don't understand is how Spring Security gets away with it. Do all of the security invocations go through the thread local?